Issue 47269: Disallow inline optimizer to optimize away function when it has been decorated with 'experimental_implements'

threebrooks opened the pull request:
Reference issue: https://github.com/tensorflow/tensorflow/issues/47269

The cause of the issue (i.e. that the 'experimental_implements' flag did not cause the PrepareCompositeFunctionsPass::ConvertTFImplements function to be called) was that the inline optimizer optimized away the function entirely, and thereby removing said flag.
Unless I misunderstand the intended functionality, when a function has been decorated with that flag, it should remain entirely untouched, since it is intended to be transformed by ConvertTFImplements.

Tfhe fix in the PR checks for the  "_implements"  attribute in the function (which is what 'experimental_implements' gets transformed into) and returns as un-optimizable.

threebrooks committed:
The inline validation accidentally deleted the experimental_implements flag during optimization. Arguably, functions with that decorator should not be optimized away in the first place

threebrooks commented:
@googlebot I signed it!

allenlavoie commented:
Please add a regression test for the behavior you're fixing. Probably fine if it's more of an integration test involving the Lite converter if that's easier; potentially someone involved with the converter should review if so.

threebrooks committed:
Add checks for it the experimental_implements was not handled, and add regression test

threebrooks commented:
@allenlavoie I added the regression test. Note that I also added checks for when a function has been declared with experimental_implements, but it was not picked up by anything in the ConvertTFAPIImplements function(s). I don't see how that shouldn't throw an error.
The regression test exercises the string version of experimental_implements, and replaces a dummy function with embedding_matmul and checks for correct output.

allenlavoie commented:
Neat. The change looks good to me, but I have limited familiarity with Lite. @abattery is probably a better reviewer, and may have more insight on the ConvertTFAPIImplements question.

abattery commented:
@allenlavoie we don't want to error out when there is no matching function with the attributes since we can just the function without fusing to run and no available fused function is not an error.

abattery commented on tensorflow/compiler/mlir/lite/transforms/prepare_composite_functions_tf.cc:
```diff
@@ -281,6 +281,9 @@ void PrepareCompositeFunctionsPass::ConvertTFImplements(FuncOp func,
     if (failed(image_warping.RewriteFunc())) {
       return signalPassFailure();
     }
+  } else {
```
We don't need to report an error when there are no matching fused functions.

threebrooks replied:
@abattery 
These checks only get executed when a function has been explicitly decorated with the "experimental_implements" keyword.
The issue I see with not throwing an error is that, because the intent is to replace on operation with another operation of equal input and output single typo in the declaration will cause TF to quietly not do anything, and leave the original op in place. As they are input and output identical, the only way to spot it would be by manual inspection of the results tflite model, i.e. by its internal structure.

abattery replied:
That is an intended behavior. I would not classify the extra manual inspection as an error. Emitting an warning is enough since we would not classify the valid cases as an error.

threebrooks replied:
@abattery  Ok, I changed the error to a warning, and have the pass not fail when this happens.

threebrooks committed:
Change the error throwing to a warning, do not have pass fail as a result

abattery commented on tensorflow/compiler/mlir/lite/transforms/prepare_composite_functions_tf.cc:
```diff
@@ -281,6 +281,8 @@ void PrepareCompositeFunctionsPass::ConvertTFImplements(FuncOp func,
       return signalPassFailure();
     }
+  } else {
+      func.emitWarning() << "Function was flagged with experimental_implements, but no handler picked it up";
```
func.emitWarning() << "with the 'xyz' attribute does not have the corresponding fused implementation found.";

Each case does have a different attribute name.

threebrooks replied:
@abattery I can change the wording, but for at least the first two cases, the attribute name is actually the same (experimental_implements), they are instead distinguished by whether the key was a plain string, or could be parsed via the attr_value.proto structure.
In terms of the last one (which internally uses the string "tf.api_implements"), I actually don't know what to use there at all, this string only appears inside tensorflow/compiler/mlir/lite/tests/prepare-composite-functions-tf.mlir, i.e. an integration test. It is not mentioned in any doc at all.

If you could let me know what explicit strings you would like to see for those 3 warnings,  I think that may be the fastest way.
Thanks!

abattery commented on tensorflow/compiler/mlir/lite/transforms/prepare_composite_functions_tf.cc:
```diff
@@ -305,6 +307,8 @@ void PrepareCompositeFunctionsPass::ConvertTFImplementsWithAttributes(
       return signalPassFailure();
     }
+  } else {
+      func.emitWarning() << "Function was flagged with experimental_implements, but no handler picked it up";
```
Ditto

abattery commented on tensorflow/compiler/mlir/lite/transforms/prepare_composite_functions_tf.cc:
```diff
@@ -324,6 +328,8 @@ void PrepareCompositeFunctionsPass::ConvertTFAPIImplements(FuncOp func,
     if (failed(ConvertKerasLSTMLayer(func, &builder)))
       return signalPassFailure();
+  } else {
+      func.emitWarning() << "Function was flagged with experimental_implements, but no handler picked it up";
```
Ditto

threebrooks commented:
@gbaned @abattery I apologize if I misunderstood the current conversation, I was actually hoping for @abattery to give me further guidance for the wording used in those warning messages (see https://github.com/tensorflow/tensorflow/pull/48856#discussion_r632675166 ). I think once I have that wording I can amend the PR and we would be able to go on.

threebrooks commented:
@gbaned @abattery @allenlavoie Is there still interest in merging this PR?

gbaned commented:
@abattery, @allenlavoie Can you please assist on above comments from @threebrooks. Thanks!

gbaned commented:
@abattery, @allenlavoie  Any update on this PR? Please. Thanks!

gbaned commented:
@abattery, @allenlavoie  Any update on this PR? Please. Thanks!

threebrooks commented:
Closing the PR from my side.
